crates/optimism/primitives/src/op_tx_type.rs

```diff
@@ -0,0 +1,197 @@
+//! newtype pattern on op_alloy_consensus::OpTxType.
+//! OpTxType that implements reth_primitives_traits::TxType.
+//! This type is required because a Compact impl is needed on the deposit tx type.
+
+use alloy_eips::eip2718::Eip2718Error;
+use alloy_primitives::{U64, U8};
+use alloy_rlp::{Decodable, Encodable};
+use bytes::BufMut;
+use core::fmt::Debug;
+use derive_more::Display;
+use op_alloy_consensus::OpTxType as AlloyOpTxType;
+use reth_primitives_traits::TxType;
+use std::convert::TryFrom;
+
+/// Wrapper type for `AlloyOpTxType` to implement `TxType` trait.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Display, Ord, Hash)]
+pub struct OpTxType(AlloyOpTxType);
+
+impl From<AlloyOpTxType> for OpTxType {
+    fn from(tx_type: AlloyOpTxType) -> Self {
+        OpTxType(tx_type)
+    }
+}
+
+impl From<OpTxType> for u8 {
+    fn from(tx_type: OpTxType) -> Self {
+        tx_type.0 as u8
+    }
+}
+
+impl From<OpTxType> for U8 {
+    fn from(tx_type: OpTxType) -> Self {
+        U8::from(u8::from(tx_type))
+    }
+}
+
+impl TryFrom<u8> for OpTxType {
+    type Error = Eip2718Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        AlloyOpTxType::try_from(value)
+            .map(OpTxType)
+            .map_err(|_| Eip2718Error::UnexpectedType(value))
+    }
+}
```

Comment:
can be replaced with `derive_more::From` and `derive_more::TryFrom` https://docs.rs/derive_more/latest/derive_more/

Reply:
I'm not sure about this one. 

`TryFrom` cannot be used on structs, only enums. If you want me to turn this into enum just because of that, I can do it, but seems counter-intuitive.

`#[into(u8)]` worked successfully. But for `U8`, I'm getting the following error:
```
error[E0277]: the trait bound `Uint<8, 1>: From<op_alloy_consensus::OpTxType>` is not satisfied
  --> crates/optimism/primitives/src/op_tx_type.rs:21:12
   |
21 | #[into(u8, U8)]
   |            ^^ the trait `From<op_alloy_consensus::OpTxType>` is not implemented for `Uint<8, 1>`
```

In short, aside from `#[into(u8)]` and `#[from(AlloyOpTxType)]`, couldn't reduce the boilerplate

Reply:
turning it into an enum doesn't make sense, fine to leave the manual impl